[FLINK-3154][API] Upgrade from Kryo 2.x to Kryo 5.x. Removed twitter …

kurtostfeld opened the pull request:
…chill companion library which is abandoned.

## What is the purpose of the change

Update the Kryo dependency. Remove the old Kryo dependency and the companion Chill dependency.

## Brief change log

- Updated Kryo dependency from 2.24.0 to the current version 5.6.2
- Removed the Twitter Chill companion library for Kryo. This library is abandoned. It depends on an old version of Kryo. It pulls in various other dependencies, that we generally do not want.
- Various other changes to accommodate the above two.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: don't know

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

kurtostfeld commented:
The following test failure is caused by a test file, external to the main git repo, that I don't have access to change. It simply needs Class<Byte> changed to Class<? extends Byte>.

```
/tmp/junit13815350649750541837/CustomKryo.java:6: error: CustomKryo is not abstract and does not override abstract method read(Kryo,Input,Class<? extends Byte>) in Serializer
public class CustomKryo extends Serializer<Byte> implements Serializable {
```

This error with table seems like a flaky test that is unrelated to the changes in this PR:

```
Jan 05 06:56:38 06:56:38.023 [ERROR]   AdaptiveJoinTest.testWithShuffleHashJoin:47 ast ==> expected: <
Jan 05 06:56:38 LogicalProject(a1=[$0], b1=[$1], c1=[$2], d1=[$3], a2=[$4], b2=[$5], c2=[$6], d2=[$7])
Jan 05 06:56:38 +- LogicalFilter(condition=[=($0, $4)])
Jan 05 06:56:38    +- LogicalJoin(condition=[true], joinType=[inner])
Jan 05 06:56:38       :- LogicalTableScan(table=[[default_catalog, default_database, T1, source: [TestTableSource(a1, b1, c1, d1)]]])
Jan 05 06:56:38       +- LogicalTableScan(table=[[default_catalog, default_database, T2, source: [TestTableSource(a2, b2, c2, d2)]]])
Jan 05 06:56:38 > but was: <
Jan 05 06:56:38 LogicalProject(a1=[$0], b1=[$1], c1=[$2], d1=[$3], a2=[$4], b2=[$5], c2=[$6], d2=[$7])
Jan 05 06:56:38 +- LogicalFilter(condition=[=($0, $4)])
Jan 05 06:56:38    +- LogicalJoin(condition=[true], joinType=[inner])
Jan 05 06:56:38       :- LogicalTableScan(table=[[default_catalog, default_database, T1]])
Jan 05 06:56:38       +- LogicalTableScan(table=[[default_catalog, default_database, T2]])
```

snuyanzin commented:
regarding table error 
>Jan 05 06:56:38 06:56:38.023 [ERROR]   AdaptiveJoinTest.testWithShuffleHashJoin:47 ast ==> expected: <


it was fixed at https://issues.apache.org/jira/browse/FLINK-37006

rebase to master should help

davidradl commented:
@kurtostfeld I see your motivation for doing this, but you mention that there will not be binary compatibility with v1 state. Do you have an idea on how v1 Kyro state would migrate to v2 state after this fix? We are trying to encourage v2 migration - and this could be a big inhibitor.

davidradl commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/ValueComparator.java:
```diff
@@ -142,12 +143,11 @@ private void checkKryoInitialized() {
             kryo.setInstantiatorStrategy(instantiatorStrategy);
 
-            this.kryo.setAsmEnabled(true);
+            // this.kryo.setAsmEnabled(true);
```
nit: I assume we want to remove this line not comment it out

kurtostfeld commented:
@davidradl great question. That is best answered by the project leaders + managers.

davidradl commented:
@kurtostfeld on your suggestion - I have raised this with on the dev list.

davidradl commented:
Reviewed by Chi on 09/01/2025 Go back to the submitter with review comments.

nicknezis commented:
> @kurtostfeld I see your motivation for doing this, but you mention that there will not be binary compatibility with v1 state. Do you have an idea on how v1 Kyro state would migrate to v2 state after this fix? We are trying to encourage v2 migration - and this could be a big inhibitor.

@davidradl Sharing this info in case it helps answer your question. @kurtostfeld can correct me if I mistate anything. The original Pull Request, which was documented in this [FLIP-317](https://cwiki.apache.org/confluence/display/FLINK/FLIP-317%3A+Upgrade+Kryo+from+2.24.0+to+5.5.0) provided code to migrate older state forward. But there was some email discussion about keeping the Kryo upgrade for Flink 2.x. Because of this, there was another merge request created that was simpler and did not include the backwards compatibility. I think either Pull Request could work, but we just need someone to decide and approve before things get too stale. This [Jira ticket](https://issues.apache.org/jira/browse/FLINK-3154) has more discussion in the comments. Also this [email discussion](https://lists.apache.org/thread/vyov944och394lql76zb3myf2g140fdw).

kurtostfeld commented:
@nicknezis The project organizers are choosing not to respond to this. I believe their strategy is to phase Kryo out and use their own serialization code instead of upgrading Kryo.

As a Flink user, I wanted to see the project fully support new versions of Java, including records, and ideally shed some super legacy libraries. I've done some tests with the Flink 2.0 master branch, and it seems to work with Java records. This did not work with 1.x versions of Flink. I checked using records in state and doing suspend/resume and can't see any problems. This satisfies my main request.

gyfora commented:
@kurtostfeld could you please rebase this on the latest master?

gyfora commented:
@kurtostfeld , I have started a mailing list discussion regarding this, would be great if you could chime in there to get this merged asap: https://lists.apache.org/thread/odhglx8tmpdt6jnorgcsvxjqjfd169x6

kurtostfeld commented:
@flinkbot run azure

kurtostfeld commented:
@flinkbot run azure

kurtostfeld commented:
@gyfora should this be retargeted to the release-2.0 branch? That branch didn't exist when this PR was created.

gyfora commented on flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/kryo/KryoSerializerUpgradeTest.java:
```diff
@@ -31,44 +31,53 @@
 
+// These tests are disabled because Flink 2.0 is doing a major Kryo 2.x -> 5.x upgrade and
+// the new Flink is not compatible with older Kryo state.
+// Restore tests when there is a Flink 2.1 that should test compatibility with Flink 2.0
```
Can we open a blocker ticket for 2.1 on JIRA to track this? Otherwise it may get easily forgotten

kurtostfeld replied:
Is this a good blocker ticket for both of the tests you flagged?
https://issues.apache.org/jira/browse/FLINK-37404

gyfora replied:
yes, thank you!

gyfora commented on flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CEPMigrationTest.java:
```diff
@@ -53,9 +54,14 @@
  * <p>For regenerating the binary snapshot file of previous versions you have to run the {@code
  * write*()} method on the corresponding Flink release-* branch.
  */
+@Ignore("Flink 2.0 is not compatible with saved state from Flink 1.x")
```
We need a blocker ticket for 2.1 for this

kurtostfeld replied:
Same question. Is this good?
https://issues.apache.org/jira/browse/FLINK-37404

gyfora replied:
looks good!

gyfora commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/utils/AvroKryoSerializerUtils.java:
```diff
@@ -106,11 +105,55 @@ public void write(Kryo kryo, Output output, Schema object) {
+            int length = collection.size();
+            output.writeVarInt(length + 1, true);
+
+            // Efficiency could be improved for cases where all elements are the same type.
```
Will this implementation still be the same or better than what we had before?

kurtostfeld replied:
see notes below.

gyfora commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/utils/AvroKryoSerializerUtils.java:
```diff
@@ -106,11 +105,55 @@ public void write(Kryo kryo, Output output, Schema object) {
+            extends CollectionSerializer<GenericData.Array> implements Serializable {
+        @Override
+        public void write(Kryo kryo, Output output, GenericData.Array collection) {
+            String schemaAsString = collection.getSchema().toString(false);
```
Can the collection be null here?

kurtostfeld replied:
I do notice that the Kryo library built-in serializers check for null, but I've added unit tests for the null scenario and I'm testing both the Flink serializer and the Kryo serializer directly and I can't create a scenario where null gets sent to that method.

gyfora reviewed (changes requested):
Added a few comments for the Avro array serializer logic, once those are addressed I think this is ready to be merged :)

gyfora commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/utils/AvroKryoSerializerUtils.java:
```diff
@@ -106,11 +105,55 @@ public void write(Kryo kryo, Output output, Schema object) {
+            Schema.Parser schemaParser = new Schema.Parser();
+            Schema schema = schemaParser.parse(schemaAsString);
+
+            int lengthPlusOne = input.readVarInt(true);
```
Why are we writing/reading length + 1 instead of length?

kurtostfeld replied:
rewrote this. please re-review.

kurtostfeld commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/utils/AvroKryoSerializerUtils.java:
```diff
@@ -106,11 +105,55 @@ public void write(Kryo kryo, Output output, Schema object) {
+            Schema.Parser schemaParser = new Schema.Parser();
+            Schema schema = schemaParser.parse(schemaAsString);
+
+            return new GenericData.Array(schema, Collections.emptyList());
```
That is correct. Look at CollectionSerializer.copy in the Kryo library and that is correct.

I believe there are some other problems with this however, that I will need to fix.

kurtostfeld replied:
I rewrote this using the Avro serialization logic. Please re-review.

gyfora commented:
@kurtostfeld , it would be good to merge this soon. The most important thing related to the Avro types in the first step is to not introduce any regressions.

kurtostfeld commented:
@gyfora I've rewritten the Avro GenericRecord Kryo serializer and the Avro GenericData.Array serializer. Avro is also a serialization library, so the Kryo serializers just basically wrap the Avro serialization. This will be more space efficient than before.

I've also added AvroKryoSerializerTests with several tests that compile and fail on master branch. These test the Avro types with both the Flink serializers wrapping Kryo and the Kryo serializer directly.

Please re-review and I will address any issues you can see.

I've added

kurtostfeld commented:
Wow, I'm excited this actually got merged! Thanks for pushing this over the finish line @gyfora !

gyfora commented:
> Wow, I'm excited this actually got merged! Thanks for pushing this over the finish line @gyfora !

Thanks for all the work @kurtostfeld :)
